internal/ethapi/api.go

```diff
@@ -549,9 +550,82 @@ type AccountResult struct {
 	StorageProof []StorageResult `json:"storageProof"`
 }
 type StorageResult struct {
-	Key   string       `json:"key"`
-	Value *hexutil.Big `json:"value"`
-	Proof []string     `json:"proof"`
+	Key     string       `json:"key"`
+	Value   *hexutil.Big `json:"value"`
+	Proof   []string     `json:"proof"`
+	Mutated bool         `json:"mutated"`
+}
+
+// Result structs for GetStateDiffProof
+type StateDiffProof struct {
+	Header   *HeaderMeta     `json:"header"`
+	Accounts []AccountResult `json:"accounts"`
+}
+type HeaderMeta struct {
+	Number    *big.Int    `json:"number"`
+	Hash      common.Hash `json:"hash"`
+	StateRoot common.Hash `json:"stateRoot"`
+	Timestamp uint64      `json:"timestamp"`
+}
+
+func (s *PublicBlockChainAPI) GetStateDiff(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (diffdb.Diff, error) {
+	_, header, err := s.b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if err != nil {
+		return nil, err
+	}
+	return s.b.GetDiff(header.Number)
```

Comment:
Should be `return s.b.GetDiff(new(big.Int).Add(header.Number, big.NewInt(1)))`

Reply:
Because `eth_getProof` returns a proof for the state *after* the given block is executed. So if the state after block `N` is our "pre-state" then we want to get the state diff for block `N+1`.

Reply:
Other option is to instead do `eth_getProof` for block `N-1` and keep this line unchanged

Reply:
If block N is our pre-state and block N+1 is our post state, then we should call it with N and N+1, no? Seems correct to me